### PR TITLE
feat(pages,server): update data in realtime

### LINKED ISSUE #30: [Pages] Update data in realtime
It should not be necessary to click on the refresh button, the pages should subscribe to the server to be notified of changes.

One of possible approach is to use [WebSocket](https://devdocs.io/dom/websockets_api).

### PR DESCRIPTION
**Pull request by heafalan**
Fixes #30

### CODE REVIEW

**Comment by heafalan** [comment]
If the pages part is fine, i will try to adapt the tests and cli parts.

**Comment by julien-f** [comment]
What's the point of this PR? What new feature does it bring?

File: packages/server/main.js [hunk 1]
```diff
@@ -70,35 +68,47 @@ const METHODS = {
...
+      wss.send(
+        format.response(request.id, result === undefined ? true : result)
+      );
 
-  try {
-    const request = parse(await getStream(ctx.req));
-    if (request.type !== "request") {
-      throw new InvalidRequest();
+      if (request.method !== "listEntries") {
+        result = METHODS["listEntries"]();
```
Comment (julien-f) [inline packages/server/main.js]: `METHODS.listEntries`, no need for the dynamic syntax.

File: packages/server/main.js [hunk 2]
```diff
@@ -70,35 +68,47 @@ const METHODS = {
...
+        clients.forEach(function(client) {
+          if (client !== wss) {
+            client.send(format.response(request.id, result));
```
Comment (julien-f) [inline packages/server/main.js]: This should not be a response, a response must correspond to a request (that's where there is `request.id`) but these clients receive a response without having sent a request.

It should be [a notification](https://github.com/JsCommunity/json-rpc-protocol#notification).

File: packages/server/main.js [hunk 3]
```diff
@@ -70,35 +68,47 @@ const METHODS = {
...
+  });
 
-    const method = METHODS[request.method];
-    if (method === undefined) {
-      throw new MethodNotFound(request.method);
+  wss.on("close", function() {
+    console.log("Closed");
+    const index = clients.indexOf(wss);
+    if (index > -1) {
+      clients.splice(index, 1);
```
Comment (julien-f) [inline packages/server/main.js]: Too complicated, use [a `Set`](https://devdocs.io/javascript/global_objects/set) instead.

File: packages/server/main.js [hunk 4]
```diff
@@ -70,35 +68,47 @@ const METHODS = {
...
+        format.error(0, new JsonRpcError(err.message, err.code, err.data))
+      );
     }
+  });
 
-    const method = METHODS[request.method];
-    if (method === undefined) {
-      throw new MethodNotFound(request.method);
+  wss.on("close", function() {
+    console.log("Closed");
```
Comment (julien-f) [inline packages/server/main.js]: No `console.log`s in final code.

File: packages/pages/src/App.js [hunk 5]
```diff
@@ -2,6 +2,13 @@ import React from "react";
 import { provideState, injectState } from "reaclette";
 import { format, parse } from "json-rpc-protocol";
 
+var websocket = new WebSocket("ws://localhost:4000");
+console.log("Connection...");
```
Comment (julien-f) [inline packages/pages/src/App.js]: Remove this log.

File: packages/pages/src/App.js [hunk 6]
```diff
@@ -2,6 +2,13 @@ import React from "react";
...
+
+websocket.onmessage = function(event) {
+  console.log(event.data);
+};
```
Comment (julien-f) [inline packages/pages/src/App.js]: Idem

File: packages/pages/src/App.js [hunk 7]
```diff
@@ -2,6 +2,13 @@ import React from "react";
 import { provideState, injectState } from "reaclette";
 import { format, parse } from "json-rpc-protocol";
 
+var websocket = new WebSocket("ws://localhost:4000");
```
Comment (julien-f) [inline packages/pages/src/App.js]: This should be stored in the component, not as a global variable.

File: packages/pages/src/App.js [hunk 8]
```diff
@@ -59,57 +66,66 @@ export default provideState({
...
-    async initialize() {
-      await this.effects.refreshEntries();
+    initialize() {
+      const { effects } = this;
+      websocket.onopen = async function() {
+        console.log("Connected to the server");
+        await effects.refreshEntries();
+      };
+      websocket.onerror = function(event) {
+        console.log(event);
```
Comment (julien-f) [inline packages/pages/src/App.js]: `console.error`.

File: packages/pages/src/App.js [hunk 9]
```diff
@@ -59,57 +66,66 @@ export default provideState({
...
-        await this.effects.refreshEntries();
-      } catch (error) {
-        console.error(error);
-      }
+        })
+      );
+      websocket.onmessage = async function(event) {
+        try {
+          await parse.result(await event.data);
+          await effects.refreshEntries();
```
Comment (julien-f) [inline packages/pages/src/App.js]: You should no longer need to refresh entries manual after making a change, the server should notify you.

File: packages/pages/src/App.js [hunk 10]
```diff
@@ -59,57 +66,66 @@ export default provideState({
...
-      }
+        })
+      );
+      websocket.onmessage = async function(evt) {
+        try {
+          await parse.result(await evt.data);
+          state.id = "";
+          state.name = "";
+          state.content = "";
+          await effects.refreshEntries();
```
Comment (julien-f) [inline packages/pages/src/App.js]: Idem.

**Review by julien-f — APPROVED** [review]

**Pull request merged**